### ISSUE TITLE
Port pre-built binaries to upstream releases

### ISSUE THREAD

**Issue by kk-hainq** [MELD-labs/fourmolu-binaries#3]
See https://github.com/fourmolu/fourmolu/issues/159. 

Our https://github.com/MELD-labs/github-artifacts doesn't seem to be affected? We can just help complete it upstream, archive this repository, then update `github-artifacts` to maintain the current flow.

**Comment by kk-hainq** [comment]
https://github.com/fourmolu/fourmolu/releases/tag/v0.6.0.0 includes a Linux binary now?

**Comment by novadev94** [comment]
`fourmolu-binaries` also builds a Darwin binary - which to be honest, I've never used and am not sure if other team members do. I still think it's valuable in a non `nix-shell` development environment because there're far too many things to install already.

I also went through their CircleCI configuration https://github.com/fourmolu/fourmolu/blame/master/.circleci/config.yml, and didn't see any mentions of adding pre-built binaries to GitHub releases... Not sure where it came from.

**Comment by kk-hainq** [comment]
If there are far too many things to install then why would we dedicate a whole repository for a single thing? How many more of these will we create? Adding any version of `fourmolu` to a Nix shell is trivial. Outside of it, `fourmolu` is bundled with HLS, or can be easily installed with `cabal`/`stack`. Like, had we forked and contributed upstream we now have the `fourmolu` team themselves maintain this stuff, instead of doing it ourselves while breaking downstream pre-commit setups with it (a separate issue, but still).

If CI didn't produce the binary then a maintainer might have built it locally which is perfectly fine (if it works that is...). I'd test that one and archive this repository if it works. At the end of the day, one of the best values of formatters is consistency, updating them frequently just to break things or cause large no-op changes makes no sense. If the next release doesn't provide a binary then let it be, we can live with `0.6.0.0` for a year no problem.

**Comment by novadev94** [comment]
Yeah, I agree; we shouldn't maintain any repositories just for hosting binaries. We can contribute to their CircleCI setup and leave the pre-built binaries in the CI artifacts if that's even needed.

To be honest, I'm leaning towards dropping the Darwin binary, as we mainly need the Linux binary to speed up our CI setup. And I also don't think any of our dev machines need that Darwin binary, as it's kinda unusable at the moment because of https://github.com/MELD-labs/github-artifacts/issues/5.

Btw, not related to this issue, but I was having trouble integrating `fourmolu` `0.6.0.0` to my `nix-shell` due to weird interactions between `fourmolu_0_6_0_0`, `Cabal`, `ghc-lib-parser` and `hls-fourmolu-plugin` (forcing `fourmolu >=0.3 && <0.6`). Would be great if I can get it done because now there're versions mismatch between Nix and non-Nix users in many of our repos...

**Comment by kk-hainq** [comment]
Yeah, I had to bump `dapp-starter`'s Nix shell yesterday while helping Duong complete an inactive PR too. Very annoying that an inactive codebase can be broken by a random remote bump, which forces the inactive code base to bump a lot of dependencies itself to bypass a minor formatting rule.

Anyhow, if there is a version mismatch you can always provide a different version with a simple `override` in Nix, like in:
https://github.com/MELD-labs/dapp-starter/blob/d04e5a4840a6b7c64ab4d3f0af3d7503a834f0ed/shell.nix#L6-L9

In this instance, we can also override `nixpkgs` itself but that would affect too many packages (that might not have been cached too). Only desired when the overridden package is key to the whole dependency tree. `fourmolu` is used as a standalone executable so locally overriding it should be better.

**Comment by novadev94** [comment]
Just tried out your config, `fourmolu -v` inside `nix-shell` still shows `0.4.0.0`:
```
[nix-shell:~/work/akamon-contracts]$ fourmolu -v
fourmolu 0.4.0.0 UNKNOWN UNKNOWN
using ghc-lib-parser 9.0.2.20211226
```

If I understand things correctly, it got overridden by `fourmolu` embedded in `hls-fourmolu-plugin`... Maybe we need to wait until this gets released then... https://github.com/haskell/haskell-language-server/pull/2854/files#diff-657e190925739b6d16d46aa6956dda0e9de823cbe251a3672c5c6e412945ce5cR26.

Btw this was the furthest I could go
```nix
let
  pkgs = import (builtins.fetchTarball "https://github.com/NixOS/nixpkgs/archive/5d12171c44af34cd0a7fc5fd3d0927f6ec42a795.tar.gz") {};

  haskellPkgs = pkgs.haskellPackages.override {
    overrides = self: super: {
      hls-fourmolu-plugin = pkgs.haskell.lib.doJailbreak super.hls-fourmolu-plugin;
      fourmolu = super.fourmolu_0_6_0_0.override {
        Cabal = self.Cabal_3_6_3_0;
        ghc-lib-parser = self.ghc-lib-parser_9_2_2_20220307;
      };
    };
  };

  devTools = [
    pkgs.git
    pkgs.gnupg
    pkgs.pre-commit
    pkgs.haskell.compiler.ghc8107
    haskellPkgs.cabal-install
    haskellPkgs.haskell-language-server
    haskellPkgs.hlint
    haskellPkgs.fourmolu
  ];

  certFile = "${pkgs.cacert}/etc/ssl/certs/ca-bundle.crt";
in
  pkgs.mkShell {
    name = "akamon-contracts";
    buildInputs =
      devTools
      ++ [
        pkgs.libsodium
        pkgs.lzma
        pkgs.pkgconfig
        pkgs.zlib
      ];
    GIT_SSL_CAINFO = certFile;
    NIX_SSL_CERT_FILE = certFile;

    shellHook = ''
      pre-commit install --install-hooks -t pre-commit -t commit-msg
    '';
  }
```

and got
```
No uhc found
*** abort because of serious configure-time warning from Cabal
error: builder for '/nix/store/ij6bh0x0dvg6iijainqwn85payymn5pb-haskell-language-server-1.6.1.1.drv' failed with exit code 1;
       last 10 log lines:
       > /nix/store/5b9jn0wcnkbvjhkgmw5bmd4zgg97j6v8-cctools-binutils-darwin-wrapper-949.0.1/bin/ld
       > No pkg-config found
       > Using runghc version 9.0.2 found on system at:
       > /nix/store/vc2q0ldb8jz49d04abalvnx849i3dxwr-ghc-9.0.2/bin/runghc
       > Using strip found on system at:
       > /nix/store/fjd0p8rk93db76l7wnk2jxh32x9yadfw-cctools-binutils-darwin-949.0.1/bin/strip
       > Using tar found on system at:
       > /nix/store/rdv54a5m4392psvpb4dmjl0bzvm5kagb-gnutar-1.34/bin/tar
       > No uhc found
       > *** abort because of serious configure-time warning from Cabal
       For full logs, run 'nix log /nix/store/ij6bh0x0dvg6iijainqwn85payymn5pb-haskell-language-server-1.6.1.1.drv'.
```

**Comment by kk-hainq** [comment]
No, the above setup should yield `fourmolu 0.6.0.0` in the Nix shell and overrides the one from HLS. Unless macOS is weird again. Why do we need something like `hls-fourmolu-plugin` if `fourmolu` is in `PATH`?

**Comment by novadev94** [comment]
Sorry it was me being dumb, `fourmolu_0_6_0_0` needs to be specified before `haskell-language-server`. Never thought that ordering matters in `buildInputs`, TIL.

About `hls-fourmolu-plugin`, it's a dependency of `haskell-language-server`, and we can't opt it out. It also installs an older version of `fourmolu`, and `nix-shell` uses whatever goes first in the `buildInputs`.

**Comment by kk-hainq** [comment]
> Never thought that ordering matters in `buildInputs`, TIL.

That's how `$PATH` works too, standard.

> About `hls-fourmolu-plugin`, it's a dependency of `haskell-language-server`, and we can't opt it out.

We can definitely opt it out. It is a library dependency of HLS to teach it how to interact with an external `fourmolu` executable. That interface hasn't changed, notice the constraint bump on `fourmolu` in `hls-fourmolu-plugin.cabal` doesn't result in any change in `Ide.Plugin.Fourmolu`. This is mostly meaningful to HLS maintainers for dependency management and house-keeping. If our use case is to add `fourmolu 0.6.0.0` to our Shell and HLS still knows how to interact with it then just add 3 extra lines to patch `fourmolu` and move on, no need to waste time tweaking a **runtime** dependency (HLS)'s library dependency that changes nothing anyway.

**Comment by novadev94** [comment]
Sorry, I think we might have misunderstood each other a bit.

> About hls-fourmolu-plugin, it's a dependency of haskell-language-server, and we can't opt it out.

was the reply for

> Why do we need something like hls-fourmolu-plugin if fourmolu is in PATH?

The Nix dependency path is: `haskell-language-server` > `hls-fourmolu-plugin` > `fourmolu` and my train of thought was: I shouldn't let Nix build 2 different derivations of `fourmolu`, and since I couldn't find a way (there should be) to remove `hls-fourmolu-plugin` out of `haskell-language-server` dependency list, I decided to "redirect" the `fourmolu` dependency to `fourmolu_0_6_0_0` - which was all completely wrong as proven.

I should have done it better. Anw, just want to clear out the probable misunderstandings.
